daemon/indexing/apollo/index.js

```diff
@@ -4,37 +4,268 @@ var search = require('../lib/search.js')
+    offers(page: Page, order: OfferOrder, filter: OfferFilter): OfferPage
+    reviews(page: Page, order: ReviewOrder, filter: ReviewFilter): ReviewPage
+  }
+
+  enum OfferStatus {
+    CREATED
+    ACCEPTED
+    DISPUTED
+    # Note: There is no "Finalized" status stored on the Offer contract on-chain.
+    #       This is computed by the indexing server.
+    FINALIZED
+  }
+
+  type Offer {
+    id: ID!
+    ipfsHash: ID!
+    listingId: ID!
+    buyer: User!
+    status: OfferStatus!
+  }
+
+  type OfferPage implements OutputPage {
+    totalCount: Int!
+    more: Boolean!
+    offers: [Offer]
+  }
+
+  type Review {
+    ipfsHash: ID!
+    reviewer: User!
+    text: String!
+    rating: Int!
+  }
+
+  type ReviewPage implements OutputPage {
+    totalCount: Int!
+    more: Boolean!
+    reviews: [Review]
+  }
+
+  # TODO: Add a status indicating if Listing is sold out.
   type Listing {
-    id: String
-    name: String
+    id: ID!
+    ipfsHash: ID!
+    seller: User!
+    title: String!
     description: String
-    price: Float
+    category: String!
+    subCategory: String!
+    price: Price!
+    offers(page: Page, order: OfferOrder, filter: OfferFilter): OfferPage
+    reviews(page: Page, order: ReviewOrder, filter: ReviewFilter): ReviewPage
+  }
+
+  type ListingPage implements OutputPage {
+    totalCount: Int!
+    more: Boolean!
+    listings: [Listing]
+  }
+
+  ######################
+  #
+  # Query input schema.
+  #
+  # Note: Some input types have a "in" prefix because GraphQL does not allow to
+  # use thw same name for both an input and output type.
+  #
+  ######################
+  enum OrderDirection {
+    ASC   # Default if no direction specified.
+    DESC
+  }
+
+  input Page {
+    num: Int!  # Page number.
+    size: Int! # Number of items per page.
+  }
+
+  input inPrice {
+    currency: String!
+    amount: String!
+  }
+
+  #
+  # ORDER
+  #
+  enum OfferOrderField {
+    CREATION_DATE
+    STATUS
+  }
+
+  enum ReviewOrderField {
+    CREATION_DATE
+  }
+
+  enum ListingOrderField {
+    RELEVANCE  # Default if no order field specified in the query.
+    PRICE
+    CREATION_DATE
+    SELLER_RATING
+  }
+
+  input OfferOrder {
+    field: OfferOrderField!
+    order: OrderDirection
+  }
+
+  input ReviewOrder {
+    field: ReviewOrderField!
+    order: OrderDirection
+  }
+
+  input ListingOrder {
+    field: ListingOrderField!
+    order: OrderDirection
+  }
+
+  #
+  # FILTERS
+  #
+
+  input OfferFilter {
+    status: OfferStatus
+  }
+
+  input ReviewFilter {
+    rating: Int
+  }
+
+  # TODO:
+  #  - Filtering definition needs more thinking. This is not flexible at all...
+  #  - Add location based filtering.
+  #  - Add fractional usage (e.g. availability) filtering.
+  input ListingFilter {
+    priceMin: inPrice
+    priceMax: inPrice
+    category: String
+    subCategory: String
+    locale: String
+    sellerAddress: String
+    buyerAddress: String
```

Comment:
Just a thought (not sure if it makes sense) to add `has_reviews`, `has_offers`

Reply:
If needed the caller can ask for the offers and reviews field. So not sure we need this ?
Leaving it out for now and east to add it later in case it turns out to be useful.